import pandas as pd
configfile: "config/config.yaml"
#GENERAL
INDIR = config["general"]["rawdir"]
SAMPLEOUT = config["general"]["sampleout"]
THREADS = config["general"]["threads"]
#GENOME AND GTF
GTF = config["genome"]["gtf"]
GENDIR = config["genome"]["dir"]
GENNAME = config["genome"]["name"]
GENEXT = config["genome"]["ext"]
GENOME = GENDIR+GENNAME+GENEXT
#FASTQ
SAMPLEEXT = config["raw_spec"]["ext"]
R12 = config["raw_spec"]["R12"]
#STAR PARAMETERS
STAR_OPT = config["star"]
#SAMTOOLS
SAMTOOLS_OPT = config["samtools"]
README_file = config["general"]["files_info"]

README = pd.read_table(README_file).set_index("Sample Name", drop=False)

README = README["Run"].to_dict()

SAMPLES = README.keys()


include:"rules/SRR.smk"
include:"rules/STAR_quant.smk"
include:"rules/samtools.smk"
include:"rules/bamcoverage.smk"

wildcard_constraints:
   accession="SRR\w+",
   prefix="[\w\-_]+"

#ALL
rule all:
	input:
		expand(SAMPLEOUT+"/mapping/BIGWIG/{sample}_{type}.bw",sample = SAMPLES,type = ["normalized_positive_strand","normalized_negative_strand","normalized"]),
		expand(SAMPLEOUT+"/mapping/counts/STAR/{sample}_counts.tsv",sample=SAMPLES),
		"logs/STARunload_Log.out"


